### ISSUE TITLE
Dynatrace will Block a VM if Dynatrace is removed before the drain-script

### ISSUE THREAD

**Issue by Espe0n** [Dynatrace/bosh-oneagent-release#52]
Dear Dynatrace-Team

We currently have the Issue, where Dynatrace will block a Deployment-Process.

## Cause:
The Issue is caused, when the uninstall-script has been called or when the drain script did not complete successfully.
In this Case Monit will try to revive the OneAgent using the `start-oneagent`-Script. However, since it does not contain an installer call, it will fail.

## Symptoms:
The affected VM will go into the “STOPPED” state. When trying to redeploy the following Error occurs:

```
Task 1234567 | 16:00:00 | Updating instance dummy: dummy/ffffffff-ffff-ffff-ffff-ffffffffffff (0) (canary) (00:02:00)
                     L Error: Action Failed get_task: Task ffffffff-ffff-ffff-ffff-ffffffffffff result: Stopping Monitored Services: Stopping services '[dynatrace-oneagent]' errored
Task 1234567 | 16:00:00 | Error: Action Failed get_task: Task ffffffff-ffff-ffff-ffff-ffffffffffff result: Stopping Monitored Services: Stopping services '[dynatrace-oneagent]' errored
```

## Workaround

You either have to force-delete the affected deployment and redeploy it afterwards, or manually ssh to the affected VM and call the pre-start-script manually.

## Possible Solution

An easy Solution would be to add a recovery-routine to the start-script.

Something likes this:

```
if ! runServiceCommand start; then
  echo "error: Could not start Dynatrace OneAgent-Service"
  echo "info: An attempt is made to repair the local Dynatrace OneAgent-Service"
  if install_dynatrace; then
     if ! runServiceCommand start; then
       echo "error: Could not repair the local Dynatrace OneAgent-Service"
       exit 1
     fi
  else
    log "error: Could not repair the local Dynatrace OneAgent-Service"
    exit 1
  fi
fi
```
Where `install_dynatrace` is the setup-function from the pre-start-script

Thanks, and best Regards
Christoph

**Comment by arthfl** [comment]
Hi Christoph,

i copied this over to our internal issue tracker and we'll take measures to make the behaviour more stable here.
Most likely incorporating a variant of your suggested fix :-)

**Comment by arthfl** [comment]
Proposed changes added in PR #53 
Once this is merged we'll create a new release :-)

**Comment by arthfl** [comment]
@Espe0n 
After some internal discussions about this whole topic, one question came up:
Was our OneAgent uninstalled manually by someone or was this some kind of weird mishap during removal of our bosh-release?

**Comment by arthfl** [comment]
@Espe0n ping :-)

If i don't get a response, we'll assume the observed behaviour was caused by a manual intervention and close this as "won't do" for now.

**Comment by Espe0n** [comment]
Hi @arthfl,
please excuse the late answer, I was not at the PC for a long time because of the holidays and am now working through the emails ...

> _Was our OneAgent uninstalled manually by someone or was this some kind of weird mishap during removal of our bosh-release?_

The Dynatrace OneAgent was not manually removed. The Issue happend during an Update of the VM using Bosh. In the Log-Files we could see, that the Drain-Script was called and successfully removed the Dynatrace OneAgent. 
However, for some reason Bosh didn't recognized this, so that the pre-start script was not called afterwards.
In the Result, the VM got stuck with the descripted Status.

Best Regard
Christoph

**Comment by arthfl** [comment]
Hi @Espe0n,

thanks for the answer :-)
Was this on a Linux or a Windows host?

**Comment by Espe0n** [comment]
Hi,
the affected host is a Linux
Ubuntu 16.04

Best regards
Christoph

**Comment by arthfl** [comment]
Ok, got you.

Added some mitigations to the Linux job, will be merged with PR #53

**Comment by arthfl** [comment]
Released v1.3.1, which fixes this: https://github.com/Dynatrace/bosh-oneagent-release/releases/tag/v1.3.1
